pkgs/development/libraries/aspell/default.nix

```diff
@@ -10,6 +10,7 @@ stdenv.mkDerivation rec {
 
   patchPhase = ''
     patch interfaces/cc/aspell.h < ${./clang.patch}
+    patch -p1 < ${./data-dirs-from-nix-profiles.patch}
```

Comment:
This needs to be an optional patch, because looking for `NIX_PROFILES` is basically an impurity. See the `if [ -z "\$ASPELL_CONF" ]; then` in the code that is removed. This is important for `aspellWithPackages`.

Reply:
For build-time purity, I agree. For run-time purity, I think it's a bit of a gray area. Taking this purity argument to the extreme would be that e.g. "bash" on NixOS would not see "git" unless you have a "bash = bashWithPackages [ git ]" thing. And I don't think we want that :-)

An alternative could be to modify the patch so that it only reads NIX_PROFILES if NIX_ASPELL_USE_PROFILES is set (or something). Then the patch could be applied unconditionally. (No recompilation needed.)

Reply:
> For build-time purity, I agree. For run-time purity, I think it's a bit of a gray area. Taking this purity argument to the extreme would be that e.g. "bash" on NixOS would not see "git" unless you have a "bash = bashWithPackages [ git ]" thing. And I don't think we want that :-)

That's an impurity that exists by default, and one which is quite useful. Yet, if you want to, you could prevent it by creating a `buildEnv` and creating wrappers setting e.g `PATH`. 

In this case, the impurity can also be useful, however it is now forced, thus requiring a wrapper unsetting `NIX_PROFILES`.

If I wanted these kind of impurities, I could have stayed with Debian or any other distro.

Reply:
This condition ensure that whenever `ASPELL_CONF` contains `data-dir` or `dict-dir`, `NIX_PROFILES` is ignored.
https://github.com/deedrah/nixpkgs/blob/f3859bfcb81a9bdefc2d536f577164418f3c7071/pkgs/development/libraries/aspell/data-dirs-from-nix-profiles.patch#L17

I have tested that `aspellWithDicts` behaves as it should with this patch.

I don't think another variable is needed, but adding new parameter for the derivation like `searchNixProfiles` for safety would be OK.